tiledb/bioimg/converters/png.py

```diff
@@ -0,0 +1,215 @@
+        self._source_cfg = self._source_ctx.config()
+        self._dest_ctx = _get_ctx(dest_ctx, dest_config)
+        self._dest_cfg = self._dest_ctx.config()
+        self._vfs = VFS(config=self._source_cfg, ctx=self._source_ctx)
+        self._vfs_fh = self._vfs.open(input_path, mode="rb")
+        self._png = Image.open(self._vfs_fh)
+
+        # Get initial metadata
+        self._metadata: Dict[str, Any] = self._png.info
+        self._metadata.update(dict(self._png.getexif()))
+
+        # Handle all different modes as RGB for consistency
+        if self._png.mode not in ["RGB", "RGBA"]:
+            self._metadata["original_mode"] = self._png.mode
+            self._png = self._png.convert("RGB")
+
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        self._vfs.close(file=self._vfs_fh)
+
+    @property
+    def source_ctx(self) -> Ctx:
+        return self._source_ctx
+
+    @property
+    def dest_ctx(self) -> Ctx:
+        return self._dest_ctx
+
+    @property
+    def logger(self) -> Optional[logging.Logger]:
+        return self._logger
+
+    @logger.setter
+    def logger(self, default_logger: logging.Logger) -> None:
+        self._logger = default_logger
+
+    @property
+    def axes(self) -> Axes:
+        axes = Axes(["X", "Y", "C"])
+        self._logger.debug(f"Reader axes: {axes}")
+        return axes
+
+    @property
+    def channels(self) -> Sequence[str]:
+        if self.webp_format is WebpInputFormat.WEBP_RGB:
+            self._logger.debug(f"Webp format: {WebpInputFormat.WEBP_RGB}")
+            return "RED", "GREEN", "BLUE"
+        self._logger.debug(f"Webp format is not: {WebpInputFormat.WEBP_RGB}")
+        color_map = {"R": "RED", "G": "GREEN", "B": "BLUE"}
+        # Use list comprehension to convert the short form to full form
+        rgb_full = [color_map[color] for color in self._png.getbands()]
+        return rgb_full
+
+    @property
+    def level_count(self) -> int:
+        level_count = 1
+        self._logger.debug(f"Level count: {level_count}")
+        return level_count
+
+    def level_dtype(self, level: int = 0) -> np.dtype:
+        dtype = np.uint8
+        self._logger.debug(f"Level {level} dtype: {dtype}")
+        return dtype
+
+    def level_shape(self, level: int = 0) -> Union[Tuple[int, ...], Any]:
+        if level != 0:
+            raise ValueError(
+                "PNG images are not pyramidal and consist of only one level"
+            )
+        # Even after converting to RGB the size is not updated to 3d from 2d
+        l_shape = (*self._png.size, 3)
+        self._logger.debug(f"Level {level} shape: {l_shape}")
+        return l_shape
+
+    @property
+    def webp_format(self) -> WebpInputFormat:
+        self._logger.debug(f"Keyframe photometric: {self._png.mode}")
+        if self._png.mode == "RGB":
+            return WebpInputFormat.WEBP_RGB
+        return WebpInputFormat.WEBP_NONE
+
+    def level_image(
```

Comment:
Unless this naming is already agreed-on / committed to, I'd suggest changing it.

In English this is ambiguous, since "level" is a noun, an adjective, and a verb. The name might mean:

* Is the image something that is level? Apparently not since the return type isn't `bool`
* Return a copy of a non-level image, for which the copy is level
* Make the image level in-place -- apparently not given the return type
* Extract one level out of a multi-image level

I'd suggest maybe calling `get_image_level`, but now I'm confused since the `level` argument of the method isn't used within the method body ...

Also, please add a docstring

Reply:
https://github.com/TileDB-Inc/TileDB-BioImaging/pull/136#discussion_r1753840251